### PR TITLE
Issue 853 aggregation zip

### PR DESCRIPTION
**Pull request by ccmoloney**
Combined PR (sorry) - also incorporates https://github.com/aodn/backlog/issues/882 - because there were inter-dependencies in the code.

Verify by running up a stack and running some test transactions.  The output should now be in ZIP file form - containing the aggregation output, a metadata XML file, and a request.xml file.

The emails sent re: jobs starting, completing and error emails should now contain details about the request - including the collection name & the subset parameters.

### CODE REVIEW

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 1]
```diff
@@ -245,38 +300,97 @@ public void run(String... args) {
...
+                zipContent.add(convertedFile.toFile());
 
                 //  URL for the status page for this job
                 String statusUrl = WpsConfig.getStatusServiceHtmlEndpoint(batchJobId);
 
+                //  Read the metadata record for the layer
+                String catalogueURL = WpsConfig.getProperty(GEONETWORK_CATALOGUE_URL_CONFIG_KEY);
+                String layerSearchField = WpsConfig.getProperty(GEONETWORK_CATALOGUE_LAYER_FIELD_CONFIG_KEY);
+                CatalogueReader catalogueReader = new CatalogueReader(catalogueURL, layerSearchField);
+                String metadataResponseXML = catalogueReader.getMetadataXML(layer);
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Longish method now - could the detail be moved to a helper method her?

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 2]
```diff
@@ -307,15 +428,22 @@ public void run(String... args) {
...
                 if (contactEmail != null) {
                     try {
                         //  Send completed job email to user
-                        emailService.sendCompletedJobEmail(contactEmail, batchJobId, statusUrl, S3Utils.getExpirationinDays(outputBucketName));
+                        emailService.sendCompletedJobEmail(
+                                contactEmail,
+                                batchJobId,
+                                statusUrl,
+                                S3Utils.getExpirationinDays(outputBucketName),
+                                EmailService.formatRequestDetail(subsetParams, collectionTitle));
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Why doesn't the email service do this there?  Its kind of complicated to have to call the EmailService to format parameters to be passed to the email service.

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 3]
```diff
@@ -408,4 +539,15 @@ private String formatPeriodString(Period period) {
...
+
+
+    private String getMetadataFilename(String collectionTitle) {
+        if(collectionTitle == null || collectionTitle.trim().length() == 0) {
+            //  Use a default filename
+            return DEFAULT_METADATA_FILENAME;
+        }
+
+        //  Replace spaces with underscores
+        return collectionTitle.replace(" ", "_") + METADATA_FILE_EXTENSION;
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: May be other unacceptable file name characters in a title?  Must be some standard code to use for that

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java [hunk 4]
```diff
@@ -102,8 +106,42 @@ public String execute() throws OGCException {
...
+                String layer = requestHelper.getLiteralInputValue(LITERAL_INPUT_IDENTIFIER_LAYER);
+                String subset = requestHelper.getLiteralInputValue(LITERAL_INPUT_IDENTIFIER_SUBSET);
+                SubsetParameters subsetParams = SubsetParameters.parse(subset);
+
+                String collectionTitle = layer;
+
+                // Lookup collection metadata to get the collection title
+                try {
+                    //  Read the metadata record for the layer
+                    String catalogueURL = WpsConfig.getProperty(GEONETWORK_CATALOGUE_URL_CONFIG_KEY);
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java]: Looks similar to code in aggregationrunner - can we pull this out into common?

File: wps-common/src/main/java/au/org/aodn/aws/geonetwork/CatalogueReader.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+        } catch (Exception e) {
+            logger.error("Could not retrieve metadata Point Of Truth URL from XML [{}]", xmlMetadataRecord, e);
+        }
+
+        return "";
+    }
+
+
+    //  The metadata response from geoserver includes the metadata record and a summary element in the form:
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/geonetwork/CatalogueReader.java]: GeoNetwork?

File: wps-common/src/main/java/au/org/aodn/aws/geoserver/client/SubsetParameters.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package au.org.emii.geoserver.client;
+package au.org.aodn.aws.geoserver.client;
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/geoserver/client/SubsetParameters.java]: We use emii everywhere else so just using aodn here is inconsistent and not sure why aws is relevant here.

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java [hunk 7]
```diff
@@ -113,4 +122,76 @@ public static String getCompletedJobEmailTemplate() {
     public static String getFailedJobEmailTemplate() {
         return String.format("%s/%s", WpsConfig.EMAIL_TEMPLATES_LOCATION, WpsConfig.FAILED_JOB_EMAIL_TEMPLATE_NAME);
     }
+
+    public static String formatRequestDetail(SubsetParameters subsetParameters, String collection) {
+
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: Would be nice to use a consistent style in java  - we don't normally put blank lines after {  - and its not done elsewhere in this PR.

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java [hunk 8]
```diff
@@ -113,4 +122,76 @@ public static String getCompletedJobEmailTemplate() {
...
+        String details = "";
+
+        if (subsetParameters != null && collection != null) {
+
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: And here

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java [hunk 9]
```diff
@@ -113,4 +122,76 @@ public static String getCompletedJobEmailTemplate() {
...
+            String spatialStr = EmailService.portalFormatSpatial(subsetParameters.getBbox());
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: Don't really need to include EmalService. here.  Its a method on this class.

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java [hunk 10]
```diff
@@ -113,4 +122,76 @@ public static String getCompletedJobEmailTemplate() {
...
+            }
+
+            return spatial;
+        }
+
+        return null;
+    }
+
+    public static String formatDate(Date date) {
+
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: Would be good to be consistent with new lines in the one class. I don't normally include them here.  Happy to put it to a vote though.

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 11]
```diff
@@ -146,30 +171,58 @@ public void run(String... args) {
...
+            }
+
             // Determine required output mime type
             String requestedMimeType = requestHelper.getRequestedMimeType("result");
-            String resultMime = requestedMimeType != null ? requestedMimeType : "application/x-netcdf";
 
-            HttpIndexReader indexReader = new HttpIndexReader(WpsConfig.getProperty(WpsConfig.GEOSERVER_CATALOGUE_ENDPOINT_URL_CONFIG_KEY));
+            logger.info("Requested mime type: " + requestedMimeType);
+
+            String resultMime = requestedMimeType != null ? requestedMimeType : DEFAULT_OUTPUT_MIME;
```
[the remark this answers is not in the document]
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Addressed in the next pull request

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 12]
```diff
@@ -227,11 +280,13 @@ public void run(String... args) {
             //  Apply connect/read timeouts
             Downloader downloader = new Downloader(downloadConnectTimeout, downloadReadTimeout);
 
+            //  Create a temp file as the destination for the aggregation
             Path outputFile = Files.createTempFile(jobDir, "agg", ".nc");
             Path convertedFile = null;
 
             long chunkSize = Long.valueOf(WpsConfig.getProperty(CHUNK_SIZE_KEY));
 
+            //  Download the and aggregate the files
```
[the remark this answers is not in the document]
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Will fix

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java (hunk 1 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: I'll have a look at what refactoring we can do

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java (hunk 2 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Done

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java (hunk 3 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Replaced with regex replace.

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java (hunk 4 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java]: Maybe once day  ;)  (I'll look at it)

File: wps-common/src/main/java/au/org/aodn/aws/geonetwork/CatalogueReader.java (hunk 5 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/geonetwork/CatalogueReader.java]: potato/potato :D  (fixed)

File: wps-common/src/main/java/au/org/aodn/aws/geoserver/client/SubsetParameters.java (hunk 6 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/geoserver/client/SubsetParameters.java]: Just makin' stuff up...  (will fix)

Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/geoserver/client/SubsetParameters.java]: (won't fix) - backlog worthy

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java (hunk 8 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: bloody gsatimos!  :)

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java (hunk 9 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: Fix

File: wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java (hunk 10 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/util/EmailService.java]: Fixed.

**Pull request merged**